docs/abi.rst

```diff
@@ -0,0 +1,916 @@
+        i = ScratchVar(TealType.uint64)
+        return For(
+            i.store(Int(0)), i.load() < array.length(), i.store(i.load() + Int(1))
+        ).Do(
+            array[i.load()].use(lambda value: Assert(value.get() > Int(5)))
+        )
+
+.. _Reference Types:
+
+Reference Types
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Many applications require the caller to provide "foreign array" values when calling the app. These are the blockchain entities (such as accounts, assets, or other applications) that the application will interact with when executing this call. In the ABI, we have **Reference Types** to describe these requirements.
+
+Definitions
+""""""""""""""""""""""""""""""""""""""""""
+
+PyTeal supports the following Reference Types:
+
+====================== ====================== ================ =======================================================================================================================================================
+PyTeal Type            ARC-4 Type             Dynamic / Static Description
+====================== ====================== ================ =======================================================================================================================================================
+:any:`abi.Account`     :code:`account`        Static           Represents an account that the current transaction can access, stored in the :any:`Txn.accounts <TxnObject.accounts>` array
+:any:`abi.Asset`       :code:`asset`          Static           Represents an asset that the current transaction can access, stored in the :any:`Txn.assets <TxnObject.assets>` array
+:any:`abi.Application` :code:`application`    Static           Represents an application that the current transaction can access, stored in the :any:`Txn.applications <TxnObject.applications>` array
+====================== ====================== ================ =======================================================================================================================================================
+
+These types all inherit from the abstract class :any:`abi.ReferenceType`.
+
+Limitations
+""""""""""""""""""""""""""""""""""""""""""
+
+Because References Types have a special meaning, they should not be directly created, and they cannot be assigned a value by a program.
+
+Additionally, Reference Types are only valid in the arguments of a method. They may not appear in a method's return value.
+
+Note that the AVM has `limitations on the maximum number of foreign references <https://developer.algorand.org/docs/get-details/parameter_tables/#smart-contract-constraints>`_ an application call transaction may contain. At the time of writing, these limits are:
+
+* Accounts: 4
+* Assets: 8
+* Applications: 8
+* Sum of Accounts, Assets, and Applications: 8
+
+Because of this, methods that have a large amount of Reference Type arguments may be impossible to call as intended at runtime.
```

Comment:
nit: just a stylistic suggestion, would you consider using `warning` or `note` in rst format to emphasize this paragraph?

Reply:
Good idea, I will wrap that in a warning